### PR TITLE
Development add test package

### PR DESCRIPTION
**Pull request by manuelserpa**
Publish and Test Packages

### CODE REVIEW

File: cmf-cli/Commands/publish/PublishCommand.cs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        /// <param name="outputDir">Destination for the dependencies package and also used for the current package</param>
+        /// <param name="repoUri">Source Location where the package dependencies should be downloaded</param>
+        /// <param name="packageId">Source Package ID</param>
+        /// <param name="packageVersion">Source Package Version</param>
+        /// <param name="loadedPackages">List of packages already processed.</param>
+        /// <param name="publishTests">True to publish test packages</param>
+        private void PublishDependenciesFromPackage(DirectoryInfo outputDir, Uri repoUri, string packageId, string packageVersion, List<string> loadedPackages, bool publishTests)
+        {
+            loadedPackages ??= new List<string>();
+            string dependencyFileName = $"{packageId}.{packageVersion}.zip";
```
Comment (lmponte) [inline cmf-cli/Commands/publish/PublishCommand.cs:39]: Shouldn't this be called `packageFileName`? We are not dealing with dependencies here.

File: cmf-cli/Utilities/GenericUtilities.cs [hunk 2]
```diff
@@ -101,6 +102,60 @@ public static void GetCurrentPresentationVersion(string source, out string versi
...
+                                dependencyFile.CopyTo(destDependencyFile);
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    throw new CliException(CliMessages.UrlsNotSupported);
+                }
+            }
```
Comment (lmponte) [inline cmf-cli/Utilities/GenericUtilities.cs:155]: if repoUri is not provided, this just returns false. Do we want to throw an exception?

File: cmf-cli/Commands/publish/PublishCommand.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            #region Output Directories Handling
+
+            DirectoryInfo packageDirectory = cmfPackage.GetFileInfo().Directory;
+            outputDir = FileSystemUtilities.GetOutputDir(cmfPackage, outputDir, true);
+            if (outputDir == null)
+            {
+                return;
+            }
+
+            DirectoryInfo packageOutputDir = FileSystemUtilities.GetPackageOutputDir(cmfPackage, packageDirectory);
```
Comment (lmponte) [inline cmf-cli/Commands/publish/PublishCommand.cs:190]: This doesn't appear to be used, except in the `finally`. Do we really need to do this?

File: cmf-cli/Enums/PackageType.cs [hunk 4]
```diff
@@ -68,6 +68,10 @@ public enum PackageType
         /// <summary>
         /// The database
```
Comment (lmponte) [inline cmf-cli/Enums/PackageType.cs:69]:
```suggestion
        /// The IoT Data
```

File: cmf-cli/Handlers/PackageType/PackageTypeHandler.cs [hunk 5]
```diff
@@ -176,6 +176,14 @@ private void GenerateDeploymentFrameworkManifest(DirectoryInfo packageOutputDir)
...
                         foreach (object property in listProperty)
                         {
+                            // if is Dependency and package type is Test, then skip the dependency on Manifest
+                            // This needs to be changed to get the Type from the dependency list.
+                            // if (property is Dependency &&
+                            //     ((Dependency)property).Type = PackageType.Test)
+                            // {
+                            //     continue;
+                            // }
+
```
Comment (lmponte) [inline cmf-cli/Handlers/PackageType/PackageTypeHandler.cs:186]: Remove commented code. Logic for test package is already encapsulated in its respective handler (it does not even generate a manifest)

**Review by lmponte — CHANGES_REQUESTED** [review]
Just some small notes and suggestions. Most important one is the mention to specific projects in a vscode config file.

File: .vscode/launch.json [hunk 6]
```diff
@@ -10,12 +10,14 @@
...
             "args": [
-                "pack",
-                "hello"
+                "publish",
+                "-r",
+                "\\\\vm-cree\\Shares\\DeliverablesPackageRepository\\development",
+                "-t"
             ],
-            "cwd": "${workspaceFolder}/cmf-cli",
+            "cwd": "C:/TFS/CREE/Tests",
```
[the remark this answers is not in the document]
Comment (manuelserpa) [reply .vscode/launch.json]: Fixed

File: cmf-cli/Commands/publish/PublishCommand.cs (hunk 1 above)
Comment (manuelserpa) [reply cmf-cli/Commands/publish/PublishCommand.cs]: Fixed

File: cmf-cli/Commands/bump/BumpCommand.cs [hunk 7]
```diff
@@ -123,6 +123,30 @@ public void Execute(CmfPackage cmfPackage, string version, string buildNr, strin
                 }
             }
 
+            if (all && cmfPackage.TestPackages.HasAny())
+            {
+                // Read all local manifests
+                CmfPackageCollection packagePathCmfPackages = packageDirectory.LoadCmfPackagesFromSubDirectories();
```
[the remark this answers is not in the document]
Comment (manuelserpa) [reply cmf-cli/Commands/bump/BumpCommand.cs]: Fixed

File: cmf-cli/Commands/publish/PublishCommand.cs (hunk 3 above)
Comment (manuelserpa) [reply cmf-cli/Commands/publish/PublishCommand.cs]: Fixed

File: cmf-cli/Enums/PackageType.cs (hunk 4 above)
Comment (manuelserpa) [reply cmf-cli/Enums/PackageType.cs]: Fixed
Also change the comment of the None item

File: cmf-cli/Handlers/PackageType/PackageTypeHandler.cs (hunk 5 above)
Comment (manuelserpa) [reply cmf-cli/Handlers/PackageType/PackageTypeHandler.cs]: Fixed

File: cmf-cli/Utilities/GenericUtilities.cs (hunk 2 above)
Comment (manuelserpa) [reply cmf-cli/Utilities/GenericUtilities.cs:155]: It was decided in the DevOps team that until other types or repositories are supported, this exception should remain

**Review by lmponte — APPROVED** [review]

**Pull request merged**